### ISSUE TITLE
Pool based allocation

### ISSUE THREAD

**Issue by vleni** [AdrienChampion/hashconsing#16]
Is there anyway to make the `HConsed` pointers allocate using some memory pool? For example the [`ArenaArc`](https://docs.rs/shared_arena/latest/shared_arena/struct.ArenaArc.html) type in [shared_arena](https://crates.io/crates/shared_arena) looks pretty compatible with `std::sync::Arc`, and it would help a lot for applications where a large number of objects needs to be allocated and deallocated.

**Comment by vleni** [comment]
For example maybe we could have
```rust
let mut factory: HConsign<Expr> = HConsign::from_arena(arena);
factory.mk(Expr::Add(1, 2)); // Calls `arena.alloc_arc(Expr::Add(1,2))`
```

**Comment by AdrienChampion** [comment]
This sounds great, but unfortunately I don't have a lot of time to spend on this...

Would you be interested in writing a PR for this?

**Comment by vleni** [comment]
> This sounds great, but unfortunately I don't have a lot of time to spend on this...
> 
> Would you be interested in writing a PR for this?

Yes. I can write a PR for it. Do you want to add `shared_arena` as a conditional dependency and use the schematic I just described?

**Comment by vleni** [comment]
There is currently a problem. The implementation of `HConsign` requires the `Weak` type, so it would look something like
```rust
pub struct HConsign<T: Hash + Eq + Clone, S = RandomState, Alloc: Allocator<T>=alloc::DefaultAllocator> {
    /// The actual hash consing table.
    table: HashMap<T, Alloc::Weak, S>,
    /// Counter for uids.
    count: u64,
}
```
`shared_arena` does not have a weak type, so I'm opening an issue [here](https://github.com/sebastiencs/shared-arena/issues/7) and hopefully we can have one.

**Comment by AdrienChampion** [comment]
I'm tempted to have two parameters for `Arc`-like refs. One is what the consign actually creates and gives back to you, and the other is what the consign actually stores.
Then we would have a trait with the constructor(s)/conversion(s) needed.

**Comment by AdrienChampion** [comment]
On that note, it seems to me like instead of having 4 or 5 or more parameters, we could just have one trait that specifies everything (types and functions) and just take that as a parameter.

Then we provide a type implementing this trait that corresponds to the current behavior and make that the default. Adding shared arena is then just a feature-gated type also implementing this trait and some helpers to easily create the associated consign.

**Comment by AdrienChampion** [comment]
What do you think?

**Comment by vleni** [comment]
> On that note, it seems to me like instead of having 4 or 5 or more parameters, we could just have one trait that specifies everything (types and functions) and just take that as a parameter. (Or maybe two traits, whatever makes the most sense, I just mean some of these parameters should be aggregated.)
> 
> Then we provide a type implementing this trait that corresponds to the current behavior and make that the default. Adding shared arena is then just a feature-gated type also implementing this trait and some helpers to easily create the associated consign.

We can do that. What do you think we should do about the Weak type? We can emulate the weak type by using two memory arenas.

Is the current solution good enough? The `Allocator<T>` type is the "one trait that specifies everything"

**Comment by vleni** [comment]
I've submitted a PR for shared_arena so the `ArenaArc` type can expose a strong count that we can use to implement bookkeeping mechanisms.

**Comment by vleni** [comment]
Update: I think `shared_arena` is no longer being maintained. I can make a fork that includes a weak pointer.
